src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -912,23 +904,17 @@ private AttributeListSyntax ParseAttributeDeclaration()
             }
 
             var attributes = _pool.AllocateSeparated<AttributeSyntax>();
-            try
+            if (attrLocation != null && attrLocation.Identifier.ToAttributeLocation() == AttributeLocation.Module)
             {
-                if (attrLocation != null && attrLocation.Identifier.ToAttributeLocation() == AttributeLocation.Module)
-                {
-                    attrLocation = CheckFeatureAvailability(attrLocation, MessageID.IDS_FeatureModuleAttrLoc);
-                }
-
-                this.ParseAttributes(attributes);
-                var closeBracket = this.EatToken(SyntaxKind.CloseBracketToken);
-                var declaration = _syntaxFactory.AttributeList(openBracket, attrLocation, attributes, closeBracket);
-
-                return declaration;
-            }
-            finally
-            {
-                _pool.Free(attributes);
+                attrLocation = CheckFeatureAvailability(attrLocation, MessageID.IDS_FeatureModuleAttrLoc);
             }
+
+            this.ParseAttributes(attributes);
+            return _syntaxFactory.AttributeList(
+                openBracket,
+                attrLocation,
+                _pool.ToListAndFree(attributes),
```

Comment:
instead of try/finally that we use so that we can just return things directly without having an intermediary cleanup step, we instead have a clean pattern we use most of hte time to just ask the pool to free the list once we are done adding to it.